pkg/server/delta/v3/server.go

```diff
@@ -44,6 +54,224 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+			} else if req.TypeUrl == "" {
+				req.TypeUrl = defaultTypeURL
+			}
+
+			typeURL := req.GetTypeUrl()
+			var state stream.StreamState
+			// Initialize the state map if we haven't already.
+			// This means that we're handling the first request for this type on the stream.
+			if s, ok := watches.deltaStreamStates[typeURL]; !ok {
+				state.ResourceVersions = make(map[string]string)
+			} else {
+				state = s
+			}
+
+			// We are in the wildcard mode if the first request of a particular type has an empty subscription list
+			var found bool
+			if state.IsWildcard, found = isWildcard[typeURL]; !found {
+				state.IsWildcard = len(req.GetResourceNamesSubscribe()) == 0
+				isWildcard[typeURL] = state.IsWildcard
+			}
+
+			s.subscribe(req.GetResourceNamesSubscribe(), state.ResourceVersions)
+			// We assume this is the first request on the stream
+			if streamNonce == 0 {
+				for r, v := range req.InitialResourceVersions {
+					state.ResourceVersions[r] = v
+				}
+			} else if streamNonce > 0 && len(req.InitialResourceVersions) > 0 {
+				return status.Errorf(codes.InvalidArgument, "InitialResourceVersions can only be set on initial stream request")
+			}
+			s.unsubscribe(req.GetResourceNamesUnsubscribe(), state.ResourceVersions)
+
+			// cancel existing watch to (re-)request a newer version
+			watch, seen := watches.deltaResponses[typeURL]
+			if !seen || watch.nonce == nonce {
+				// We must signal goroutine termination to prevent a race between the cancel closing the watch
+				// and the producer closing the watch.`	`
+				if terminate, exists := watches.deltaTerminations[typeURL]; exists {
+					close(terminate)
+				}
+				if watch.cancel != nil {
+					watch.cancel()
+				}
+				watch.responses, watch.cancel = s.cache.CreateDeltaWatch(req, &state)
+
+				// a go-routine. Golang does not allow selecting over a dynamic set of channels.
+				terminate := make(chan struct{})
+				watches.deltaTerminations[typeURL] = terminate
+				go func() {
+					select {
+					case resp, more := <-watch.responses:
+						if more {
+							watches.deltaMuxedResponses <- resp
+						} else {
+							// Check again if the watch is cancelled.
+							select {
+							case <-terminate: // do nothing
+							default:
+								// We cannot close the responses channel since it can be closed twice.
+								// Instead we send a fake error response.
+								watches.deltaMuxedResponses <- deltaErrorResponse
+							}
+						}
+						break
+					case <-terminate:
+						break
+					}
+				}()
+			}
+		}
+	}
+}
+
 func (s *server) DeltaStreamHandler(str stream.DeltaStream, typeURL string) error {
-	return nil
+	// a channel for receiving incoming delta requests
+	reqCh := make(chan *discovery.DeltaDiscoveryRequest)
+
+	go func() {
+		for {
+			select {
+			case <-str.Context().Done():
+				close(reqCh)
+				return
+			default:
+				req, err := str.Recv()
+				if err != nil {
+					close(reqCh)
+					return
+				}
+
+				reqCh <- req
+			}
+		}
+	}()
```

Comment:
Why do we need this extra goroutine? I can imagine it not being great for large deployments to require yet another goroutine per stream.

At the very least let's document why this is done

Reply:
Well go routines are pretty light weight, but I think this was originally written in a go routine from the SOTW stuff because we need to concurrently handle incoming requests since that function is actually what kicks off the `processDelta` method. If we blocked there on the for select we'd never process those incoming requests